### PR TITLE
Handle missing previous read marker event

### PR DESCRIPTION
**Pull request by Fizzadar**
### Pull Request Checklist

Closes: https://github.com/matrix-org/synapse/issues/11738

Signed off by Nick @ Beeper (@fizzadar).

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by Fizzadar** [comment]
No idea why the test timed out, seems unrelated to the change here.

File: synapse/handlers/read_marker.py [hunk 1]
```diff
@@ -46,12 +47,21 @@ async def received_client_read_marker(
...
-                    event_id, existing_read_marker["event_id"]
-                )
+                try:
+                    old_event_ordering = await self.store.get_event_ordering(
+                        existing_read_marker["event_id"]
+                    )
+                except SynapseError:
+                    # Old event no longer exists, assume new is ahead. This may
+                    # happen if the old event was removed due to retention.
+                    pass
```
Comment (clokep) [inline synapse/handlers/read_marker.py:61]: Would it be possible to add a test for this?

Comment (Fizzadar) [reply synapse/handlers/read_marker.py:61]: Yep! https://github.com/matrix-org/synapse/pull/15464/commits/cbfc04efaf2df0f41dc1e93c4dd112c9ebb9f54b

File: changelog.d/15464.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Don't fail moving read marker if current set to a missing event. Contributed by Nick @ Beeper (@fizzadar).
```
Comment (clokep) [inline changelog.d/15464.bugfix:1]: I think this is description might help server admins understand better what this is fixing? What do you think?

```suggestion
Fix a long-standing bug where setting the read marker could fail when using message retention. Contributed by Nick @ Beeper (@fizzadar).
```

File: tests/rest/client/test_read_marker.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+from synapse.api.constants import EventTypes
+from synapse.rest import admin
+from synapse.rest.client import login, read_marker, register, room
+from synapse.server import HomeServer
+from synapse.util import Clock
+
+from tests import unittest
+
+one_hour_ms = 3600000
+one_day_ms = one_hour_ms * 24
```
Comment (clokep) [inline tests/rest/client/test_read_marker.py:26]:
```suggestion
ONE_HOUR_MS = 3600000
ONE_DAY_MS = ONE_HOUR_MS * 24
```

(I think you copied this from `tests/rest/client/test_retation.py`, but no reason to propagate that.)

File: tests/rest/client/test_read_marker.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        # Send a second message (retention will not remove the latest event ever)
+        send_message()
+        # And then advance so retention rules remove the first event (where the marker is)
+        self.reactor.advance(one_day_ms * 2 / 1000)
+
+        event = self.get_success(self.store.get_event(event_id_1, allow_none=True))
+        assert event is None
+
+        self.store.get_event_ordering.invalidate_all()
```
Comment (clokep) [inline tests/rest/client/test_read_marker.py:134]: Is this a bug somewhere that we don't invalidate this cache when retention runs?

File: tests/rest/client/test_read_marker.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
+# Copyright 2022 Beeper
```
[the remark this answers is not in the document]
Comment (Fizzadar) [reply tests/rest/client/test_read_marker.py:1]: 😆  my brain is clearly stuck in 2022!

File: tests/rest/client/test_read_marker.py (hunk 4 above)
Comment (clokep) [reply tests/rest/client/test_read_marker.py:134]: #15609 seems to imply that this should be invalidated.

Comment (clokep) [reply tests/rest/client/test_read_marker.py:134]: I guess let's just add a TODO comment for now, maybe pointing to that PR?

Comment (clokep) [reply tests/rest/client/test_read_marker.py:134]:
```suggestion
        # TODO See https://github.com/matrix-org/synapse/issues/13476
        self.store.get_event_ordering.invalidate_all()
```

**Comment by clokep** [comment]
(@Fizzadar If you enable edits by maintainer I can accept the couple of remaining suggestions. 👍 )

**Comment by Fizzadar** [comment]
@clokep apparently not possible! https://github.com/orgs/community/discussions/5634 Was wondering where the button was!

File: tests/rest/client/test_read_marker.py (hunk 3 above)
Comment (Fizzadar) [reply tests/rest/client/test_read_marker.py]: https://github.com/matrix-org/synapse/pull/15464/commits/71e135349f950a663e25c36caa047f5decdd6c7a

**Review by clokep — APPROVED** [review]

**Pull request merged**